PR: https://github.com/odoo/odoo/pull/210299

From: 1c7d80a10b5d7db1c4163166bf52b3f3c77044ba
From: Mohammad Abdulmoneim (abdu)
Date: 2026-01-21 06:25:15

Breaking data model changes scores: del:11 + add:0, change matches:
-    scrap_ids = fields.One2many('stock.scrap', 'production_id', 'Scraps')
-    scrap_ids = fields.One2many('stock.scrap', 'workorder_id')
-    scrap_count = fields.Integer(compute='_compute_scrap_move_count', string='Scrap Move')
-    _inherit = 'stock.scrap'
-    production_id = fields.Many2one(
-    workorder_id = fields.Many2one(
-    product_is_kit = fields.Boolean(related='product_id.is_kits')
-    product_template = fields.Many2one(related='product_id.product_tmpl_id')
-    bom_id = fields.Many2one(

Total Changes: 203

[IMP] stock,*: remove scrap model

This commit removes `stock.scrap` model without affecting its
functionality from the user perspective.

Previous to this commit, when a user wants to scrap some product,
a record of `stock.scrap` is created which is a scrap order
containing info about the product, quantity, scrap reason, etc.
When a scrap order is confirmed, it generates a `stock.move` and
confirms it so that the quantity move is reflected in the quants.
That's why `stock.scrap` model itself provides no intrinsic value
and is dispensable.

This commit removes the model completely and relies only on
 `stock.move` to achieve the same functionality. The menu item
"Scrap" now redirects to a list of stock.moves that
has the new field `is_scrap` = true. If the user creates a new record
from the list, a draft `stock.move` is created initially. When it's
confirmed, it genrates its `stock.move.line` and its state is
changed to be done.

The only fields that are added to `stock.move` to fully replace
`stock.scrap` are `scrap_reason_tag_ids` and
`should_replenish_scrapped`.

This commit also adds a new action to scrap a lot/serial number from its
form view directly.

It's also now not possible to scrap a kit due to a limitation in
replacing `stock.scrap` with `stock.move`.

Task-4606260

closes odoo/odoo#210299

Related: odoo/enterprise#85706
Related: odoo/documentation#15776
Related: odoo/upgrade#7693
Signed-off-by: Arnold Moyaux (arm) <arm@odoo.com>

================================= pseudo patch: =================================

--- a/addons/mrp/models/__init__.py
+++ b/addons/mrp/models/__init__.py
@@ -19,7 +19,6 @@ from . import stock_picking
 from . import stock_lot
 from . import stock_reference
 from . import stock_rule
-from . import stock_scrap
 from . import stock_warehouse
 from . import stock_quant
 from . import stock_replenish_mixin

--- a/addons/mrp/models/mrp_production.py
+++ b/addons/mrp/models/mrp_production.py
@@ -204,11 +204,11 @@ class MrpProduction(models.Model):
         'stock.move', 'raw_material_production_id', 'Components',
         compute='_compute_move_raw_ids', store=True, readonly=False,
         copy=False,
-        domain=[('location_dest_usage', '!=', 'inventory')])
+        domain=[('is_scrap', '=', False)])
     move_finished_ids = fields.One2many(
         'stock.move', 'production_id', 'Finished Products', readonly=False,
         compute='_compute_move_finished_ids', store=True, copy=False,
-        domain=[('location_dest_usage', '!=', 'inventory')])
+        domain=[('is_scrap', '=', False)])
     # technical field: inverse field for `stock.move.raw_material_production_id`
     all_move_raw_ids = fields.One2many('stock.move', 'raw_material_production_id')
     # technical field: inverse field for `stock.move.production_id`
@@ -247,7 +247,6 @@ class MrpProduction(models.Model):
         help='If checked, when the previous move of the move (which was generated by a next procurement) is cancelled or split, the move generated by this move will too')
     delay_alert_date = fields.Datetime('Delay Alert Date', compute='_compute_delay_alert_date', search='_search_delay_alert_date')
     json_popover = fields.Char('JSON data for the popover widget', compute='_compute_json_popover')
-    scrap_ids = fields.One2many('stock.scrap', 'production_id', 'Scraps')
     scrap_count = fields.Integer(compute='_compute_scrap_move_count', string='Scrap Move')
     unbuild_ids = fields.One2many('mrp.unbuild', 'mo_id', 'Unbuilds')
     unbuild_count = fields.Integer(compute='_compute_unbuild_count', string='Number of Unbuilds')
@@ -704,8 +703,10 @@ class MrpProduction(models.Model):
         return True
 
     def _compute_scrap_move_count(self):
-        data = self.env['stock.scrap']._read_group([('production_id', 'in', self.ids)], ['production_id'], ['__count'])
-        count_data = {production.id: count for production, count in data}
+        scrapped_finished_data = self.env['stock.move']._read_group([('production_id', 'in', self.ids), ('is_scrap', '=', True)], ['production_id'], ['__count'])
+        scrapped_component_data = self.env['stock.move']._read_group([('raw_material_production_id', 'in', self.ids), ('is_scrap', '=', True)], ['raw_material_production_id'], ['__count'])
+        count_data = {production.id: count for production, count in scrapped_finished_data}
+        count_data.update({production.id: count_data.get(production.id, 0) + count for production, count in scrapped_component_data})
         for production in self:
             production.scrap_count = count_data.get(production.id, 0)
 
@@ -2369,27 +2370,36 @@ class MrpProduction(models.Model):
     def do_unreserve(self):
         (self.move_finished_ids | self.move_raw_ids).filtered(lambda x: x.state not in ('done', 'cancel'))._do_unreserve()
 
-    def button_scrap(self):
+    def action_scrap(self):
         self.ensure_one()
         return {
             'name': _('Scrap Products'),
             'view_mode': 'form',
-            'res_model': 'stock.scrap',
-            'views': [[self.env.ref('stock.stock_scrap_form_view2').id, 'form']],
+            'res_model': 'stock.move',
+            'views': [[self.env.ref('stock.view_scrap_move_form').id, 'form']],
             'type': 'ir.actions.act_window',
-            'context': {'default_production_id': self.id,
-                        'product_ids': (self.move_raw_ids.filtered(lambda x: x.state not in ('done', 'cancel')) | self.move_finished_ids.filtered(lambda x: x.state == 'done')).mapped('product_id').ids,
-                        'default_company_id': self.company_id.id
-                        },
+            'context': {
+                'from_order': True,
+                'product_ids': (self.move_raw_ids.filtered(lambda x: x.state not in ('done', 'cancel')) | self.move_finished_ids.filtered(lambda x: x.state == 'done')).mapped('product_id').ids,
+                'default_is_scrap': True,
+                'default_production_id': self.id,
+                'default_company_id': self.company_id.id,
+                'default_location_id': self.location_dest_id.id if self.state == 'done' else self.location_src_id.id,
+                'default_location_dest_id': self.company_id.scrap_location_id.id,
+            },
             'target': 'new',
         }
 
     def action_see_move_scrap(self):
         self.ensure_one()
-        action = self.env["ir.actions.actions"]._for_xml_id("stock.action_stock_scrap")
-        action['domain'] = [('production_id', '=', self.id)]
-        action['context'] = dict(self.env.context, default_origin=self.name)
-        return action
+        return {
+            'name': _('Scraps'),
+            'type': 'ir.actions.act_window',
+            'res_model': 'stock.move',
+            'views': [(self.env.ref('stock.view_scrap_move_list').id, 'list'), (self.env.ref('stock.view_scrap_move_form').id, 'form')],
+            'domain': ['&', '|', ('raw_material_production_id', '=', self.id), ('production_id', '=', self.id), ('is_scrap', '=', True)],
+            'context': dict(self.env.context, default_origin=self.name, create=False),
+        }
 
     def action_view_reception_report(self):
         action = self.env["ir.actions.actions"]._for_xml_id("mrp.mrp_reception_action")

--- a/addons/mrp/models/mrp_workorder.py
+++ b/addons/mrp/models/mrp_workorder.py
@@ -129,8 +129,6 @@ class MrpWorkorder(models.Model):
     # Technical field to store the cost_mode of a workorder in case it is changed on the operation_id later on.
     # This field should only be changed once at MO confirmation and should reflect the cost_mode of the operation_id.
 
-    scrap_ids = fields.One2many('stock.scrap', 'workorder_id')
-    scrap_count = fields.Integer(compute='_compute_scrap_move_count', string='Scrap Move')
     production_date = fields.Datetime('Production Date', compute='_compute_production_date', store=True)
     json_popover = fields.Char('Popover Data JSON', compute='_compute_json_popover')
     show_json_popover = fields.Boolean('Show Popover?', compute='_compute_json_popover')
@@ -426,12 +424,6 @@ class MrpWorkorder(models.Model):
             else:
                 order.is_user_working = False
 
-    def _compute_scrap_move_count(self):
-        data = self.env['stock.scrap']._read_group([('workorder_id', 'in', self.ids)], ['workorder_id'], ['__count'])
-        count_data = {workorder.id: count for workorder, count in data}
-        for workorder in self:
-            workorder.scrap_count = count_data.get(workorder.id, 0)
-
     @api.onchange('operation_id')
     def _onchange_operation_id(self):
         if self.operation_id:
@@ -762,27 +754,6 @@ class MrpWorkorder(models.Model):
             production._plan_workorders(replan=True)
         return True
 
-    def button_scrap(self):
-        self.ensure_one()
-        return {
-            'name': _('Scrap Products'),
-            'view_mode': 'form',
-            'res_model': 'stock.scrap',
-            'views': [(self.env.ref('stock.stock_scrap_form_view2').id, 'form')],
-            'type': 'ir.actions.act_window',
-            'context': {'default_company_id': self.production_id.company_id.id,
-                        'default_workorder_id': self.id,
-                        'default_production_id': self.production_id.id,
-                        'product_ids': (self.production_id.move_raw_ids.filtered(lambda x: x.state not in ('done', 'cancel')) | self.production_id.move_finished_ids.filtered(lambda x: x.state == 'done')).mapped('product_id').ids},
-            'target': 'new',
-        }
-
-    def action_see_move_scrap(self):
-        self.ensure_one()
-        action = self.env["ir.actions.actions"]._for_xml_id("stock.action_stock_scrap")
-        action['domain'] = [('workorder_id', '=', self.id)]
-        return action
-
     def action_open_wizard(self):
         self.ensure_one()
         action = self.env["ir.actions.actions"]._for_xml_id("mrp.mrp_workorder_mrp_production_form")

--- a/addons/mrp/models/stock_move.py
+++ b/addons/mrp/models/stock_move.py
@@ -15,7 +15,7 @@ class StockMove(models.Model):
         if self.env.context.get('default_raw_material_production_id') or self.env.context.get('default_production_id'):
             production_id = self.env['mrp.production'].browse(self.env.context.get('default_raw_material_production_id') or self.env.context.get('default_production_id'))
 
-            if production_id.state not in ('draft', 'cancel'):
+            if production_id.state not in ('draft', 'cancel') and 'state' not in defaults:
                 if production_id.state != 'done':
                     defaults['state'] = 'draft'
                 else:
@@ -101,12 +101,13 @@ class StockMove(models.Model):
     def _compute_location_dest_id(self):
         ids_to_super = set()
         for move in self:
+            if move.is_scrap or (not move.production_id and not move.raw_material_production_id):
+                ids_to_super.add(move.id)
+                continue
             if move.production_id:
                 move.location_dest_id = move.production_id.location_dest_id
             elif move.raw_material_production_id:
                 move.location_dest_id = move.product_id.with_company(move.company_id).property_stock_production.id
-            else:
-                ids_to_super.add(move.id)
         return super(StockMove, self.browse(ids_to_super))._compute_location_dest_id()
 
     @api.depends('bom_line_id')
@@ -369,7 +370,7 @@ class StockMove(models.Model):
         moves_ids_to_unlink = OrderedSet()
         phantom_moves_vals_list = []
         for move in self:
-            if (not move.picking_type_id and not (self.env.context.get('is_scrap') or self.env.context.get('skip_picking_assignation'))) or (move.production_id and move.production_id.product_id == move.product_id):
+            if (not move.picking_type_id and not (move.is_scrap or self.env.context.get('skip_picking_assignation'))) or (move.production_id and move.production_id.product_id == move.product_id):
                 moves_ids_to_return.add(move.id)
                 continue
             bom = self.env['mrp.bom'].sudo()._bom_find(move.product_id, company_id=move.company_id.id, bom_type='phantom')[move.product_id]
@@ -638,3 +639,26 @@ class StockMove(models.Model):
                         for move in self):
             res = 'assigned'
         return res
+
+    def _action_replenish(self, values=False):
+        self.ensure_one()
+        values = values or {}
+        group = self.production_id.production_group_id or self.raw_material_production_id.production_group_id
+        if group:
+            values.update({
+                'production_group_id': group.id,
+            })
+        super()._action_replenish(values)
+
+    def _prepare_scrap_move_vals(self):
+        self.ensure_one()
+        res = super()._prepare_scrap_move_vals()
+        res.update({
+            'production_id': False,
+            'raw_material_production_id': False,
+        })
+        if self.product_id in self.production_id.move_finished_ids.product_id:
+            res['production_id'] = self.production_id.id
+        else:
+            res['raw_material_production_id'] = self.production_id.id
+        return res

--- a/addons/mrp/models/stock_scrap.py
+++ b/None
@@ -1,99 +0,0 @@
-# -*- coding: utf-8 -*-
-# Part of Odoo. See LICENSE file for full copyright and licensing details.
-
-from odoo import _, api, fields, models
-
-
-class StockScrap(models.Model):
-    _inherit = 'stock.scrap'
-
-    production_id = fields.Many2one(
-        'mrp.production', 'Manufacturing Order',
-        index='btree_not_null',
-        check_company=True)
-    workorder_id = fields.Many2one(
-        'mrp.workorder', 'Work Order',
-        index='btree_not_null',
-        check_company=True) # Not to restrict or prefer quants, but informative
-    product_is_kit = fields.Boolean(related='product_id.is_kits')
-    product_template = fields.Many2one(related='product_id.product_tmpl_id')
-    bom_id = fields.Many2one(
-        'mrp.bom', 'Kit',
-        domain="[('type', '=', 'phantom'), '|', ('product_id', '=', product_id), '&', ('product_id', '=', False), ('product_tmpl_id', '=', product_template)]",
-        check_company=True)
-
-    @api.depends('workorder_id', 'production_id')
-    def _compute_location_id(self):
-        remaining_scrap = self.browse()
-
-        for scrap in self:
-            if scrap.production_id:
-                if scrap.production_id.state != 'done':
-                    scrap.location_id = scrap.production_id.location_src_id.id
-                else:
-                    scrap.location_id = scrap.production_id.location_dest_id.id
-            elif scrap.workorder_id:
-                scrap.location_id = scrap.workorder_id.production_id.location_src_id.id
-            else:
-                remaining_scrap |= scrap
-
-        res = super(StockScrap, remaining_scrap)._compute_location_id()
-        return res
-
-    def _prepare_move_values(self):
-        vals = super(StockScrap, self)._prepare_move_values()
-        if self.production_id:
-            vals['origin'] = vals['origin'] or self.production_id.name
-            if self.product_id in self.production_id.move_finished_ids.mapped('product_id'):
-                vals.update({'production_id': self.production_id.id})
-            else:
-                vals.update({'raw_material_production_id': self.production_id.id})
-        return vals
-
-    @api.onchange('lot_id')
-    def _onchange_serial_number(self):
-        if self.product_id.tracking == 'serial' and self.lot_id:
-            if self.production_id:
-                message, recommended_location = self.env['stock.quant'].sudo()._check_serial_number(self.product_id,
-                                                                                                    self.lot_id,
-                                                                                                    self.company_id,
-                                                                                                    self.location_id,
-                                                                                                    self.production_id.location_dest_id)
-                if message:
-                    if recommended_location:
-                        self.location_id = recommended_location
-                    return {'warning': {'title': _('Warning'), 'message': message}}
-            else:
-                return super()._onchange_serial_number()
-
-    @api.onchange('product_id')
-    def _onchange_product_id(self):
-        if self.product_is_kit:
-            self.bom_id = self.env['mrp.bom']._bom_find(self.product_id, company_id=self.company_id.id, bom_type='phantom')[self.product_id]
-        else:
-            self.bom_id = False
-
-    @api.depends('move_ids', 'move_ids.move_line_ids.quantity', 'product_id')
-    def _compute_scrap_qty(self):
-        self.scrap_qty = 1
-        for scrap in self:
-            if not scrap.bom_id:
-                return super(StockScrap, scrap)._compute_scrap_qty()
-            if scrap.move_ids:
-                filters = {
-                    'incoming_moves': lambda m: True,
-                    'outgoing_moves': lambda m: False
-                }
-                scrap.scrap_qty = scrap.move_ids._compute_kit_quantities(scrap.product_id, scrap.scrap_qty, scrap.bom_id, filters)
-
-    def _should_check_available_qty(self):
-        return super()._should_check_available_qty() or self.product_is_kit
-
-    def do_replenish(self, values=False):
-        self.ensure_one()
-        values = values or {}
-        if self.production_id and self.production_id.production_group_id:
-            values.update({
-                'production_group_id': self.production_id.production_group_id.id,
-            })
-        super().do_replenish(values)
